Lib/pathlib/__init__.py

```diff
@@ -209,7 +209,14 @@ def __hash__(self):
         try:
             return self._hash
         except AttributeError:
-            self._hash = hash(self._str_normcase)
+            hash_data = self._tail
+            if self._drv or self._root:
+                hash_data = [self._drv + self._root] + self._tail
+            elif self._tail and self.parser.splitdrive(self._tail[0])[0]:
+                hash_data = ['.'] + self._tail
```

Comment:
```suggestion            
            if self._drv or self._root:
                hash_data = [self._drv + self._root] + self._tail
            elif self._tail and self.parser.splitdrive(self._tail[0])[0]:
                hash_data = ['.'] + self._tail
            else:
                hash_data = self._tail
```
Is slightly faster. 

Or
```suggestion
            hash_data = self._tail
            if self._drv or self._root:
                hash_data = [self._drv + self._root] + hash_data 
            elif self._tail and self.parser.splitdrive(hash_data [0])[0]:
                hash_data = ['.'] + hash_data 
```
(although I find this less readable)

For both suggestions: I would prefer to keep the code close to how `_str_normcase` was implemented, to make it easier to keep the code aligned when the `_str_normcase` implementation changes. Maybe even add a test that `hash(p)` equals `hash(p._str_normcase)` (if not already there)